src/peep/peep.c

```diff
@@ -3421,7 +3439,585 @@ static void peep_update_fixing(int steps, rct_peep* peep){
+		uint16 stationX = (stationPosition & 0xFF) * 32;
+		uint16 stationY = (stationPosition >> 8) * 32;
+
+		rct_map_element *mapElement = map_get_track_element_at(stationX, stationY, stationZ);
+		if (mapElement == NULL) {
+			log_error("Couldn't find map_element");
+			return false;
+		}
+
+		int direction = map_element_get_direction(mapElement);
+		rct_xy16 offset = RCT2_ADDRESS(0x992A3C, rct_xy16)[direction * 2];
+
+		stationX += 16 + offset.x;
+		if (offset.x == 0) {
+			stationX = peep->destination_x;
+		}
+
+		stationY += 16 + offset.y;
+		if (offset.y == 0) {
+			stationY = peep->destination_y;
+		}
+
+		peep->destination_x = stationX;
+		peep->destination_y = stationY;
+		peep->destination_tolerence = 2;
+	}
+
+	invalidate_sprite_2((rct_sprite *) peep);
+	if (!peep_update_action(&x, &y, &tmp_distance, peep)) {
+		return true;
+	}
+
+	sprite_move(x, y, peep->z, (rct_sprite *) peep);
+	invalidate_sprite_2((rct_sprite *) peep);
+
+	return false;
+}
+
+/**
+ * rct2: 0x006C11F5
+ */
+static bool peep_update_fixing_sub_state_8(bool firstRun, rct_peep *peep) {
+	sint16 tmp_x, tmp_y, tmp_xy_distance;
+
+	if (!firstRun) {
+		peep->sprite_direction = peep->var_78 << 3;
+		peep->action = PEEP_ACTION_STAFF_CHECKBOARD;
+		peep->action_frame = 0;
+		peep->action_sprite_image_offset = 0;
+
+		sub_693B58(peep);
+		invalidate_sprite_2((rct_sprite *) peep);
+	}
+
+	if (peep->action == PEEP_ACTION_NONE_2) {
+		return true;
+	}
+
+	peep_update_action(&tmp_x, &tmp_y, &tmp_xy_distance, peep);
+
+	return false;
+}
+
+/**
+ * rct2: 0x006C1239
+ */
+static bool peep_update_fixing_sub_state_9(bool firstRun, rct_peep *peep, rct_ride *ride) {
+	sint16 x, y, tmp_xy_distance;
+
+	if (!firstRun) {
+		if (ride_type_has_flag(ride->type, RIDE_TYPE_FLAG_3 | RIDE_TYPE_FLAG_15)) {
+			return true;
+		}
+
+		uint16 stationPosition = ride->station_starts[peep->current_ride_station];
+		if (stationPosition == 0xFFFF) {
+			return true;
+		}
+
+		uint8 stationZ = ride->station_heights[peep->current_ride_station];
+
+		rct_xy_element input;
+		input.x = (stationPosition & 0xFF) * 32;
+		input.y = (stationPosition >> 8) * 32;
+		input.element = map_get_track_element_at_from_ride(input.x, input.y, stationZ, peep->current_ride);
+
+		track_begin_end trackBeginEnd;
+		while (track_block_get_previous(input.x, input.y, input.element, &trackBeginEnd)) {
+			uint8 trackType = trackBeginEnd.begin_element->properties.track.type;
+			if (trackType == 2 || trackType == 3 || trackType == 1) {
+				input.x = trackBeginEnd.begin_x;
+				input.y = trackBeginEnd.begin_y;
+				input.element = trackBeginEnd.begin_element;
+
+				RCT2_GLOBAL(0xF43914, uint32) = trackBeginEnd.begin_element->type & 3;
+				continue;
+			}
+
+			break;
+		}
+
+		// loc_6C12ED:
+		uint8 direction = RCT2_GLOBAL(0xF43914, uint32);
+		uint16 destinationX = input.x + TileDirectionDelta[direction].x + 16;
+		uint16 destinationY = input.y + TileDirectionDelta[direction].y + 16;
```

Comment:
The existing code assumed that you had walked onto the next track piece this new code does not assume this. You shouldn't do the `+ TileDirectionDelta[direction].x/y`